### PR TITLE
Refactor index validity-checking code into a new file 

### PR DESCRIPTION
**Pull request by jayz22**
### What

Mostly just refactoring/clean-ups. Along the way fixed a couple of things. See comments. 
See commit log for details. 
(Note: the first commit d174f64 says "Bugfix" but it was not actually a bug. Just forgot to edit the commit message.)

### Why

[TODO: Why this change is being made. Include any context required to understand the why.]

### Known limitations

[TODO or N/A]

### CODE REVIEW

File: stellar-contract-env-host/src/host.rs [hunk 1]
```diff
@@ -1161,26 +1153,11 @@ impl CheckedEnv for Host {
...
-        if start > end {
-            return Err(self.err_status_msg(
-                ScHostFnErrorCode::InputArgsInvalid,
-                "start greater than end",
-            ));
-        }
+        let start = self.u32_from_rawval_input("start", start)?;
+        let end = self.u32_from_rawval_input("end", end)?;
         let vnew = self.visit_obj(v, move |hv: &HostVec| {
-            if start > hv.len() {
```
Comment (jayz22) [inline stellar-contract-env-host/src/host.rs:1173]: This was a potential bug. Should be `start >= hv.len()`

File: stellar-contract-env-host/src/host.rs [hunk 2]
```diff
@@ -1719,20 +1683,11 @@ impl CheckedEnv for Host {
     }
 
     fn binary_slice(&self, b: Object, start: RawVal, end: RawVal) -> Result<Object, HostError> {
-        let start = self.usize_from_rawval_u32_input("start", start)?;
-        let end = self.usize_from_rawval_u32_input("end", end)?;
+        let start = self.u32_from_rawval_input("start", start)?;
+        let end = self.u32_from_rawval_input("end", end)?;
         let vnew = self.visit_obj(b, move |hv: &Vec<u8>| {
-            if start > hv.len() {
```
Comment (jayz22) [inline stellar-contract-env-host/src/host.rs:1725]: This was a bug. Would've caused a slice out of bound panic if passing in `hv.len()`. 
Should be `start >= hv.len()`

File: stellar-contract-env-host/src/host/validity.rs [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+                "start index out of bound",
+            ));
+        }
+        if end as usize > bound {
+            return Err(self.err_status_msg(
+                ScHostObjErrorCode::VecIndexOutOfBound,
+                "end index out of bound",
+            ));
+        }
+        // we don't check start <= end, as [start..end] is ok in that case
```
Comment (sisuresh) [inline stellar-contract-env-host/src/host/validity.rs:51]: Is this right? Doesn't `[start..end]` where `start > end` result in a panic?

Comment (jayz22) [reply stellar-contract-env-host/src/host/validity.rs]: Thanks for the catch. You are right it does cause a panic. The im-rc had additional internal logic that just returns empty Vector if `start > end`, which tricked me to thinking that was ok in general. So glad you caught this. Fixed.

**Comment by jayz22** [comment]
Ready to be merged. Fixed the comment and reconciled with upstream renaming changes.

**Review by sisuresh — APPROVED** [review]

**Pull request merged**